### ISSUE TITLE
Error: Meteor code must always run within a Fiber. Try wrapping callbacks that you pass to non-Meteor libraries with Meteor.bindEnvironment.

### ISSUE THREAD

**Issue by solo999** [VeliovGroup/Meteor-logger-mongo#1]
When logger-mongo package was added, I got above error (just logger-console package work fine)

**Comment by dr-dimitru** [comment]
@solo999 could you please provide full error trace and .meteor/packages

**Comment by dr-dimitru** [comment]
@solo999 also do you updated to the last available version? (last release was yesterday)

**Comment by solo999** [comment]
I think it was the previous version. I'm gonna try with new one soon...

Trace is here:
Error: Meteor code must always run within a Fiber. Try wrapping callbacks that you pass to non-Meteor libraries with Meteor.bindEnvironment.
    at Object.Meteor._nodeCodeMustBeInFiber (packages/meteor/dynamics_nodejs.js:9:1)
    at [object Object]._.extend.get (packages/meteor/dynamics_nodejs.js:21:1)
    at Object.DDP.randomStream (packages/ddp/random_stream.js:69:1)
    at [object Object].self._makeNewID (packages/mongo/collection.js:72:1)
    at [object Object].Mongo.Collection.(anonymous function) (packages/mongo/collection.js:488:1)
    at [object Object].Mongo.Collection.(anonymous function) [as insert](packages/aldeed:collection2/collection2.js:159:1)
    at Object.emitter (packages/ostrio:loggermongo/ostrio:loggermongo.coffee:7:25)
    at Logger.logit (packages/ostrio:logger/ostrio:logger.coffee:51:14)
    at Logger.debug (packages/ostrio:logger/ostrio:logger.coffee:120:6)
    at Socket.<anonymous> (app/server/tcp/tcpserver.js:76:16)

**Comment by dr-dimitru** [comment]
@solo999 yes, please try to run after update to latest version, and let me know how it works. 
Also could you please provide code of module initialisation and sample where you log something. Or this error comes out even before logging something?

**Comment by solo999** [comment]
Here it is:

``` javascript
    //Init app
    this.init(function() {

        this_server.server = net.createServer(function(connection) {

            connection.device = new Device(this_server.getProtocol(), connection, this_server);

            //When raw data comes from the tracker  
            connection.on('data', function(data) {
                Meteor.log.debug('data arrived: '+data.toString(), ['tcp']);
                connection.device.emit('data', data);
            });

            //When data sent to the tracker     
            connection.on('send_data', function(msg) {
                console.log('data has sent: '+msg.toString(), ['tcp']);     
            });

            //Error handling on connection  
            connection.on('error', function(err) {
                console.log('connection error of tracker ' + connection.device.uid + ' :'+err.toString(), ['tcp']);
                connection.device.logout();
            });

            //When tracker closed the connection
            connection.on('end', function() {
                console.log('connection end of tracker ' + connection.device.uid, ['tcp']);
                connection.device.logout();
            });

            //When connection timed out on inactivity
            connection.on('timeout', function() {           
                console.log('connection timeout of tracker ' + connection.device.uid, ['tcp']);
                connection.device.logout();
            });         

            this_server.devices.push(connection);

            callback(connection.device, connection);

        }).listen(opts.port);

    });
```

**Comment by solo999** [comment]
and simple setup:

``` javascript
Meteor.log.rule('Mongo', {
    enable: true,
    filter: ['*'],
    client: true, 
    server: true 
});
```

I think I can easy get rid of error, by wrapping Meteor.log.debug in 

``` javascript
Fiber(function() {  
  Meteor.log.debug();
}).run();
```

But it would be strange solution...

**Comment by dr-dimitru** [comment]
@solo999 
Where `callback` comes from? 
And what happened when you remove line with `callback`?

Also you need something like this, to enable `MongoLogger`:

``` javascript
Meteor.log.rule('Mongo', 
{
  enable: true,
  filter: ['*'], /* Filters: 'ERROR', 'FATAL', 'WARN', 'DEBUG', 'INFO', '*' */
  client: false, /* This allows to call, but not execute on Client */
  server: true   /* Calls from client will be executed on Server */
});
```

**Comment by solo999** [comment]
callback has nothing to do with this error.
error happened on data arrived event...

**Comment by dr-dimitru** [comment]
@solo999 in Meteor there is no need to use `Fiber` directly as it is built-in, use `Meteor.bindEnvironment`, like:

``` javascript
var bound = Meteor.bindEnvironment(function(callback){
   callback()
});
```

and then just:

``` javascript
bound(function(){
   Meteor.log.debug();
});
```

**Comment by dr-dimitru** [comment]
@solo999 But I'm pretty sure Meteor want's you to wrap this line: `callback(connection.device, connection);` could you please try?

**Comment by solo999** [comment]
Yes. Correct. But Meteor.bindEnvironment is also strange solution...
As far as I understand the problem is in Collection.find as it async and require fiber...

Ok. I'll try, but I had the same problem every time when I use Collection.find within tcp server code... 
By the way: error come only when I use logger-mongo and not logger-console. I suppose logger-mongo driver also use Collection.find and that's where the problem come from...

**Comment by dr-dimitru** [comment]
@solo999 I've wrapped Collection insert to `bindEnvironment` on my end, could you roll back to initial state when you submit this issue, and try [v0.0.5](https://github.com/VeliovGroup/Meteor-logger-mongo/tree/dev)?

To do so:

``` shell
mkdir loggermongo 
cd loggermongo
git clone git@github.com:VeliovGroup/Meteor-logger-mongo.git
checkout dev

cd # to your project's folder
mrt link-package # full path to folder with package/loggermongo
meteor update
meteor
```

**Comment by solo999** [comment]
I have tried your solution. It works like a charm. Thank you...

**Comment by dr-dimitru** [comment]
Merged in: 927038e
